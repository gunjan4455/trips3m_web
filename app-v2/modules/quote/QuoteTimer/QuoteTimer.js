import React, { PureComponent } from 'react';
import PropTypes from 'prop-types';

import { secondsToDhms } from 'helpers/DateTime';
import { capitalizeFirstLetter } from 'app/utils/parsers';
import { ORGANISM_VIEWED } from 'actions/segmentEvents';
import styles from './quoteTimer.cm.scss';

const TIME_INTERVAL = 1000;

class QuoteTimer extends PureComponent {

  state = { totalTimeInSeconds: this.props.totalTimeInSeconds };

  componentDidMount() {
    const { trackSegment, totalTimeInSeconds } = this.props;
    const timeObject = secondsToDhms(totalTimeInSeconds);
    let cta;
    Object.keys(timeObject).map((key) => {
      let rt = Math.floor(timeObject[key]);
      rt = rt < 10 ? `0${rt}` : rt;
      cta = cta ? `${cta}_${rt}` : `Timer_${rt}`;
    });
    trackSegment({ event: ORGANISM_VIEWED, section: 'Urgency Timer', cta });
    this.timer = setInterval(() => this.tick(), TIME_INTERVAL);
  }

  tick = () => {
    const { totalTimeInSeconds } = this.state;
    if (totalTimeInSeconds <= 0) {
      clearInterval(this.timer);
      return;
    }
    return this.setState({ totalTimeInSeconds: totalTimeInSeconds - 1 });
  };

  componentWillUnmount() {
    clearInterval(this.timer);
  }


  renderTimeBlock = (remainingTime, timeUnit) => {
    const rt = Math.floor(remainingTime);
    return (
      <div className="flexFull text-center pr15 pl15 ">
        <p className="f16 fw9">{rt < 10 ? `0${rt}` : rt}</p>
        <span className="f10">{capitalizeFirstLetter(timeUnit)}</span>
      </div>
    );
  };

  render() {
    const { totalTimeInSeconds } = this.state;
    if (totalTimeInSeconds < 0) {
      return null;
    }
    const timeObject = secondsToDhms(totalTimeInSeconds);
    return (
      <div className="at_QuoteTimerBanner relative pb24 sbcw">
        <div className={`pt15 pb32 ${styles.timerBanner}`}>
          <h4 className="text-center f14 fw4 pl24 pr24 mb15">Hurry! Book your quote before prices increase.<br />Quote prices expire in:</h4>
        </div>
        <div className={` ${styles.countDown} at_QuoteTimerSection radius6 flex sbcw pt8 pb8 pl15 pr15 absolute l0 r0 b0`}>
        {
          Object.keys(timeObject).map((key) => this.renderTimeBlock(timeObject[key], key))
        }
        </div>
      </div>
    );
  }
}

QuoteTimer.propTypes = {
  totalTimeInSeconds: PropTypes.number.isRequired,
  trackSegment: PropTypes.func.isRequired
};

QuoteTimer.defaultProps = {
  totalTimeInSeconds: -1
};

export default QuoteTimer;

