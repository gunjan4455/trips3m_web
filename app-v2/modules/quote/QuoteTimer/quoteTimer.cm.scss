:local{
  .timerBanner{
    background:#fac22e;
  }
  .countDown{
    margin: 0 auto;
    width: 249px;
    box-shadow: 2px 2px 14px 2px rgba(0, 0, 0, 0.1);
    -webkit-box-shadow: 2px 2px 14px 2px rgba(0, 0, 0, 0.1);
    -moz-box-shadow: 2px 2px 14px 2px rgba(0, 0, 0, 0.1)  ;
    span{
      opacity: 0.56;
    }
    >div{
      border-right:1px solid #efefef;
      &:last-child{
        border:0;
      }
    }
  }
}