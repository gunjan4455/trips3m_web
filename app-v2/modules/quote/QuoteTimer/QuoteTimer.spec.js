import React from 'react';
import Adapter from 'enzyme-adapter-react-16';
import {shallow, configure} from 'enzyme';

import QuoteTimer from './QuoteTimer.js';

configure({ adapter: new Adapter()});

const  DEFAULT_PROPS={
	totalTimeInSeconds: 2*24*3600
};

const shallowWrapper = (props) => {
	return shallow(<QuoteTimer {...props} />);
}

describe('CountDownTimer component', ()=> {
	let props = {...DEFAULT_PROPS};

	beforeEach(()=> {
		props = {...DEFAULT_PROPS};
		jest.useFakeTimers();
	})

	afterEach(()=> {
		jest.clearAllTimers();
		jest.useRealTimers();	
	});


	it('should render correctly', ()=> {
		const wrapper = shallowWrapper(props);
		expect(wrapper).toMatchSnapshot();
	})

	it('should tick correctly', ()=> {
		const wrapper = shallowWrapper(props);
		jest.advanceTimersByTime(3000);
		expect(wrapper.state("totalTimeInSeconds")).toEqual(2*24*3600 - 3);
	});

	it('should finish correctly', ()=> {
		props = {
			totalTimeInSeconds: 5
		}
		const wrapper = shallowWrapper(props);
		jest.useFakeTimers();
		jest.runAllTimers();
		expect(wrapper.state("totalTimeInSeconds")).toEqual(0);
	}); 
});
