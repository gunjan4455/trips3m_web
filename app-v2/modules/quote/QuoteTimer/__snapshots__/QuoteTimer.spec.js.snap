// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CountDownTimer component should render correctly 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <QuoteTimer
    totalTimeInSeconds={172800}
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "checkPropTypes": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <h4
          className="text-center fw4 "
        >
          Hurry! Book your quote before prices increase. Quote prices expire in:
        </h4>,
        <div
          className=" countDown  radius6 flex sbcw pt8 pb8 pl15 pr15"
        >
          <div
            className="text-center pr15 pl15 "
          >
            <p
              className="f24 fw9"
            >
              02
            </p>
            <span
              className="f10"
            >
              Days
            </span>
          </div>
          <div
            className="text-center pr15 pl15 "
          >
            <p
              className="f24 fw9"
            >
              00
            </p>
            <span
              className="f10"
            >
              Hrs
            </span>
          </div>
          <div
            className="text-center pr15 pl15 "
          >
            <p
              className="f24 fw9"
            >
              00
            </p>
            <span
              className="f10"
            >
              Mins
            </span>
          </div>
          <div
            className="text-center pr15 pl15 "
          >
            <p
              className="f24 fw9"
            >
              00
            </p>
            <span
              className="f10"
            >
              Secs
            </span>
          </div>
        </div>,
      ],
      "className": "relative pt15 timerBanner",
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": "Hurry! Book your quote before prices increase. Quote prices expire in:",
          "className": "text-center fw4 ",
        },
        "ref": null,
        "rendered": "Hurry! Book your quote before prices increase. Quote prices expire in:",
        "type": "h4",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": Array [
            <div
              className="text-center pr15 pl15 "
            >
              <p
                className="f24 fw9"
              >
                02
              </p>
              <span
                className="f10"
              >
                Days
              </span>
            </div>,
            <div
              className="text-center pr15 pl15 "
            >
              <p
                className="f24 fw9"
              >
                00
              </p>
              <span
                className="f10"
              >
                Hrs
              </span>
            </div>,
            <div
              className="text-center pr15 pl15 "
            >
              <p
                className="f24 fw9"
              >
                00
              </p>
              <span
                className="f10"
              >
                Mins
              </span>
            </div>,
            <div
              className="text-center pr15 pl15 "
            >
              <p
                className="f24 fw9"
              >
                00
              </p>
              <span
                className="f10"
              >
                Secs
              </span>
            </div>,
          ],
          "className": " countDown  radius6 flex sbcw pt8 pb8 pl15 pr15",
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": Array [
                <p
                  className="f24 fw9"
                >
                  02
                </p>,
                <span
                  className="f10"
                >
                  Days
                </span>,
              ],
              "className": "text-center pr15 pl15 ",
            },
            "ref": null,
            "rendered": Array [
              Object {
                "instance": null,
                "key": undefined,
                "nodeType": "host",
                "props": Object {
                  "children": "02",
                  "className": "f24 fw9",
                },
                "ref": null,
                "rendered": "02",
                "type": "p",
              },
              Object {
                "instance": null,
                "key": undefined,
                "nodeType": "host",
                "props": Object {
                  "children": "Days",
                  "className": "f10",
                },
                "ref": null,
                "rendered": "Days",
                "type": "span",
              },
            ],
            "type": "div",
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": Array [
                <p
                  className="f24 fw9"
                >
                  00
                </p>,
                <span
                  className="f10"
                >
                  Hrs
                </span>,
              ],
              "className": "text-center pr15 pl15 ",
            },
            "ref": null,
            "rendered": Array [
              Object {
                "instance": null,
                "key": undefined,
                "nodeType": "host",
                "props": Object {
                  "children": "00",
                  "className": "f24 fw9",
                },
                "ref": null,
                "rendered": "00",
                "type": "p",
              },
              Object {
                "instance": null,
                "key": undefined,
                "nodeType": "host",
                "props": Object {
                  "children": "Hrs",
                  "className": "f10",
                },
                "ref": null,
                "rendered": "Hrs",
                "type": "span",
              },
            ],
            "type": "div",
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": Array [
                <p
                  className="f24 fw9"
                >
                  00
                </p>,
                <span
                  className="f10"
                >
                  Mins
                </span>,
              ],
              "className": "text-center pr15 pl15 ",
            },
            "ref": null,
            "rendered": Array [
              Object {
                "instance": null,
                "key": undefined,
                "nodeType": "host",
                "props": Object {
                  "children": "00",
                  "className": "f24 fw9",
                },
                "ref": null,
                "rendered": "00",
                "type": "p",
              },
              Object {
                "instance": null,
                "key": undefined,
                "nodeType": "host",
                "props": Object {
                  "children": "Mins",
                  "className": "f10",
                },
                "ref": null,
                "rendered": "Mins",
                "type": "span",
              },
            ],
            "type": "div",
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": Array [
                <p
                  className="f24 fw9"
                >
                  00
                </p>,
                <span
                  className="f10"
                >
                  Secs
                </span>,
              ],
              "className": "text-center pr15 pl15 ",
            },
            "ref": null,
            "rendered": Array [
              Object {
                "instance": null,
                "key": undefined,
                "nodeType": "host",
                "props": Object {
                  "children": "00",
                  "className": "f24 fw9",
                },
                "ref": null,
                "rendered": "00",
                "type": "p",
              },
              Object {
                "instance": null,
                "key": undefined,
                "nodeType": "host",
                "props": Object {
                  "children": "Secs",
                  "className": "f10",
                },
                "ref": null,
                "rendered": "Secs",
                "type": "span",
              },
            ],
            "type": "div",
          },
        ],
        "type": "div",
      },
    ],
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <h4
            className="text-center fw4 "
          >
            Hurry! Book your quote before prices increase. Quote prices expire in:
          </h4>,
          <div
            className=" countDown  radius6 flex sbcw pt8 pb8 pl15 pr15"
          >
            <div
              className="text-center pr15 pl15 "
            >
              <p
                className="f24 fw9"
              >
                02
              </p>
              <span
                className="f10"
              >
                Days
              </span>
            </div>
            <div
              className="text-center pr15 pl15 "
            >
              <p
                className="f24 fw9"
              >
                00
              </p>
              <span
                className="f10"
              >
                Hrs
              </span>
            </div>
            <div
              className="text-center pr15 pl15 "
            >
              <p
                className="f24 fw9"
              >
                00
              </p>
              <span
                className="f10"
              >
                Mins
              </span>
            </div>
            <div
              className="text-center pr15 pl15 "
            >
              <p
                className="f24 fw9"
              >
                00
              </p>
              <span
                className="f10"
              >
                Secs
              </span>
            </div>
          </div>,
        ],
        "className": "relative pt15 timerBanner",
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": "Hurry! Book your quote before prices increase. Quote prices expire in:",
            "className": "text-center fw4 ",
          },
          "ref": null,
          "rendered": "Hurry! Book your quote before prices increase. Quote prices expire in:",
          "type": "h4",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": Array [
              <div
                className="text-center pr15 pl15 "
              >
                <p
                  className="f24 fw9"
                >
                  02
                </p>
                <span
                  className="f10"
                >
                  Days
                </span>
              </div>,
              <div
                className="text-center pr15 pl15 "
              >
                <p
                  className="f24 fw9"
                >
                  00
                </p>
                <span
                  className="f10"
                >
                  Hrs
                </span>
              </div>,
              <div
                className="text-center pr15 pl15 "
              >
                <p
                  className="f24 fw9"
                >
                  00
                </p>
                <span
                  className="f10"
                >
                  Mins
                </span>
              </div>,
              <div
                className="text-center pr15 pl15 "
              >
                <p
                  className="f24 fw9"
                >
                  00
                </p>
                <span
                  className="f10"
                >
                  Secs
                </span>
              </div>,
            ],
            "className": " countDown  radius6 flex sbcw pt8 pb8 pl15 pr15",
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": Array [
                  <p
                    className="f24 fw9"
                  >
                    02
                  </p>,
                  <span
                    className="f10"
                  >
                    Days
                  </span>,
                ],
                "className": "text-center pr15 pl15 ",
              },
              "ref": null,
              "rendered": Array [
                Object {
                  "instance": null,
                  "key": undefined,
                  "nodeType": "host",
                  "props": Object {
                    "children": "02",
                    "className": "f24 fw9",
                  },
                  "ref": null,
                  "rendered": "02",
                  "type": "p",
                },
                Object {
                  "instance": null,
                  "key": undefined,
                  "nodeType": "host",
                  "props": Object {
                    "children": "Days",
                    "className": "f10",
                  },
                  "ref": null,
                  "rendered": "Days",
                  "type": "span",
                },
              ],
              "type": "div",
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": Array [
                  <p
                    className="f24 fw9"
                  >
                    00
                  </p>,
                  <span
                    className="f10"
                  >
                    Hrs
                  </span>,
                ],
                "className": "text-center pr15 pl15 ",
              },
              "ref": null,
              "rendered": Array [
                Object {
                  "instance": null,
                  "key": undefined,
                  "nodeType": "host",
                  "props": Object {
                    "children": "00",
                    "className": "f24 fw9",
                  },
                  "ref": null,
                  "rendered": "00",
                  "type": "p",
                },
                Object {
                  "instance": null,
                  "key": undefined,
                  "nodeType": "host",
                  "props": Object {
                    "children": "Hrs",
                    "className": "f10",
                  },
                  "ref": null,
                  "rendered": "Hrs",
                  "type": "span",
                },
              ],
              "type": "div",
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": Array [
                  <p
                    className="f24 fw9"
                  >
                    00
                  </p>,
                  <span
                    className="f10"
                  >
                    Mins
                  </span>,
                ],
                "className": "text-center pr15 pl15 ",
              },
              "ref": null,
              "rendered": Array [
                Object {
                  "instance": null,
                  "key": undefined,
                  "nodeType": "host",
                  "props": Object {
                    "children": "00",
                    "className": "f24 fw9",
                  },
                  "ref": null,
                  "rendered": "00",
                  "type": "p",
                },
                Object {
                  "instance": null,
                  "key": undefined,
                  "nodeType": "host",
                  "props": Object {
                    "children": "Mins",
                    "className": "f10",
                  },
                  "ref": null,
                  "rendered": "Mins",
                  "type": "span",
                },
              ],
              "type": "div",
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": Array [
                  <p
                    className="f24 fw9"
                  >
                    00
                  </p>,
                  <span
                    className="f10"
                  >
                    Secs
                  </span>,
                ],
                "className": "text-center pr15 pl15 ",
              },
              "ref": null,
              "rendered": Array [
                Object {
                  "instance": null,
                  "key": undefined,
                  "nodeType": "host",
                  "props": Object {
                    "children": "00",
                    "className": "f24 fw9",
                  },
                  "ref": null,
                  "rendered": "00",
                  "type": "p",
                },
                Object {
                  "instance": null,
                  "key": undefined,
                  "nodeType": "host",
                  "props": Object {
                    "children": "Secs",
                    "className": "f10",
                  },
                  "ref": null,
                  "rendered": "Secs",
                  "type": "span",
                },
              ],
              "type": "div",
            },
          ],
          "type": "div",
        },
      ],
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "legacyContextMode": "parent",
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getChildContext": Object {
            "calledByRenderer": false,
          },
          "getDerivedStateFromError": false,
          "getDerivedStateFromProps": Object {
            "hasShouldComponentUpdateBug": false,
          },
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
    Symbol(enzyme.__providerValues__): undefined,
  },
  Symbol(enzyme.__providerValues__): Map {},
  Symbol(enzyme.__childContext__): null,
}
`;
